sound/soc/sof/intel/mtl.c

```diff
@@ -581,6 +584,9 @@ static void mtl_ipc_dump(struct snd_sof_dev *sdev)
 
 static int mtl_dsp_disable_interrupts(struct snd_sof_dev *sdev)
 {
+	if (sdev->dspless_mode_selected)
+		return 0;
+
```

Comment:
we should have this test inside mtl_disable_ipc_interrupts() for consistency with the existing code in hda-dsp.c

Alternate proposal:

````diff
diff --git a/sound/soc/sof/intel/mtl.c b/sound/soc/sof/intel/mtl.c
index fe9b6671a194..7f72502e86cc 100644
--- a/sound/soc/sof/intel/mtl.c
+++ b/sound/soc/sof/intel/mtl.c
@@ -60,6 +60,9 @@ bool mtl_dsp_check_ipc_irq(struct snd_sof_dev *sdev)
        u32 irq_status;
        u32 hfintipptr;
 
+       if (sdev->dspless_mode_selected)
+               return false;
+
        /* read Interrupt IP Pointer */
        hfintipptr = snd_sof_dsp_read(sdev, HDA_DSP_BAR, MTL_HFINTIPPTR) & MTL_HFINTIPPTR_PTR_MASK;
        irq_status = snd_sof_dsp_read(sdev, HDA_DSP_BAR, hfintipptr + MTL_DSP_IRQSTS);
@@ -120,6 +123,9 @@ void mtl_enable_ipc_interrupts(struct snd_sof_dev *sdev)
        struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
        const struct sof_intel_dsp_desc *chip = hda->desc;
 
+       if (sdev->dspless_mode_selected)
+               return;
+
        /* enable IPC DONE and BUSY interrupts */
        snd_sof_dsp_update_bits(sdev, HDA_DSP_BAR, chip->ipc_ctl,
                                MTL_DSP_REG_HFIPCXCTL_BUSY | MTL_DSP_REG_HFIPCXCTL_DONE,
@@ -131,6 +137,9 @@ void mtl_disable_ipc_interrupts(struct snd_sof_dev *sdev)
        struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
        const struct sof_intel_dsp_desc *chip = hda->desc;
 
+       if (sdev->dspless_mode_selected)
+               return;
+
        /* disable IPC DONE and BUSY interrupts */
        snd_sof_dsp_update_bits(sdev, HDA_DSP_BAR, chip->ipc_ctl,

````
[mtl.diff.txt](https://github.com/thesofproject/linux/files/11159326/mtl.diff.txt)

Reply:
@plbossart I'm still having page fault with this call stack. I need to handle same thing in mtl_enable_sdw_irq()

```
[   72.947570] Call Trace:
[   72.947576]  <TASK>
[   72.947582]  snd_sof_dsp_update_bits+0x4d/0x80 [snd_sof]
[   72.947614]  mtl_enable_sdw_irq+0x35/0x190 [snd_sof_intel_hda_common]
[   72.947645]  mtl_dsp_disable_interrupts+0x14/0x60 [snd_sof_intel_hda_common]
[   72.947670]  hda_suspend+0x4c/0x150 [snd_sof_intel_hda_common]
[   72.947696]  ? __lock_acquire.constprop.0+0x117/0x690
[   72.947707]  hda_dsp_runtime_suspend+0x4c/0xe0 [snd_sof_intel_hda_common]
[   72.947729]  sof_suspend.isra.0+0xd8/0x2d0 [snd_sof]
[   72.947749]  pci_pm_runtime_suspend+0x5f/0x1b0
[   72.947757]  ? __pfx_pci_pm_runtime_suspend+0x10/0x10
```

Reply:
mtl_enable_interrupts() needs that too, i will push my update soon

```
[   73.348878] RIP: 0010:mtl_enable_interrupts+0x291/0x360 [snd_sof_intel_hda_common]
[   73.348921] Code: e8 d4 d3 0a e7 e9 cf fe ff ff 48 8b 33 45 89 e0 b9 40 11 00 00 48 c7 c2 90 8e ba c0 48 c7 c7 28 5e bb c0 e8 b1 d3 0a e7 eb a1 <44> 8b 26 e9 b5 fd ff ff 44 8b 2e e9 9a fe ff ff 44 8b 26 eb 8a 48
...
[   73.348954] Call Trace:
[   73.348959]  <TASK>
[   73.348964]  hda_suspend+0x4c/0x150 [snd_sof_intel_hda_common]
[   73.348984]  ? __lock_acquire.constprop.0+0x117/0x690
[   73.348992]  hda_dsp_runtime_suspend+0x4c/0xe0 [snd_sof_intel_hda_common]
[   73.349011]  sof_suspend.isra.0+0xd8/0x2d0 [snd_sof]
[   73.349047]  pci_pm_runtime_suspend+0x5f/0x1b0
[   73.349055]  ? __pfx_pci_pm_runtime_suspend+0x10/0x10
```

Reply:
thanks @fredoh9 

I think it's because we we HDA_DSP_BAR in mtl_enable_sdw_irq(). That's not initialized in dspless mode, only HDA_DSP_HDA_BAR is.

It's the same problem in mtl_enable_interrupts().

But I don't think there's any reason why this is required, IIRC the SoundWire bits are shadowed in the host and DSP BAR, so we could very well use the host one.

let me look into this, this could be a mistake in the code.

Reply:
@plbossart your suggestion is same as my last update, can you double check and approve this?